### DISCUSSION TITLE
NPC Tactics discussion

### DISCUSSION THREAD

**Discussion by I-am-Erk** [CleverRaven/Cataclysm-DDA#69683] [Ideas]
I have recently redone a lot of the NPC combat AI and I hope to do a lot more. I am running into the problem though that I'm not all that great at tactics, and don't play the game as much as I used to. I'd really like to crowdsource some of the planning and ideas here, and get some thoughts on what other behaviours NPCs could have. Please read on before posting though because I do need to outline some specifics.

1. For the moment, I'm only working on close range combat - basically, fighting within 5 or 6 tiles. Outside the close area, combat needs a whole different AI and I'm not up to programming that yet.
2. I really want to focus on "big picture" behaviours for now, I don't want to program a routine for "this is how to take on a mi-go" but I do like routines for "what to do against a small number of enemies that are faster and tougher than you"
3. I would prefer ideas for tactics that don't require the player to be around. That's not as important though because I do plan to have non-player-ally NPCs pick a Leader and treat them the same way that other NPCs treat the player

The AI that I am working on currently does a bunch of things. It estimates how tough the assessing NPC (let's call them `Assessor`) is, how tough its allies are, how tough its enemies are, and how outnumbered the allied team is. It also looks at how many enemies are crowding in on Assessor when they're using a ranged weapon. They have currently got two behaviours based on this: reposition and run away. Using this data they will:
- choose a place to reposition/run to. First they'll look for a player-defined Retreat Zone. If there's no retreat zone, they check if they're friends with the player, and if so, consider whether or not they should move closer to the player for protection. If neither of these are true, they'll run in the direction that they think has the fewest enemies.
- If they're just repositioning, they'll usually only move a couple tiles. If they're running, they'll do so for up to 20 turns.
- During the reposition/run process, they'll keep reassessing their situation to see if they have successfully managed to reduce the danger level around them. If the threat around them isn't dropping, they'll consider cancelling their reposition/flee. This is so that if for example they run around a corner into a larger group of zombies, they might cancel running. This is a new behaviour and I'm still testing it.

In the short term, I am hoping to add:
- Assessor looks at their allies to see if anyone is really struggling against their own enemies. If Assessor is still strong and Struggler is weak, they may choose to go closer towards that ally, ideally causing enemies to split their attention to Assessor. Note that Struggler doesn't ask for help, this is Assessor deciding they need it, so the player may also be deemed Struggler using the same math.

From here I'm at a bit of a loss. I want to make sure I break down my AI processes to make sense for what future tactics I want to add, but I am not really sure what would be good directions. I've already fixed (afaict) the things that were driving me the most crazy but I also know they can be much, much better. I'd love some opinions and ideas for where to go next.

Some things that have already come up:
- perhaps I could add a routine for NPCs to spot a particular high priority target, like a necromancer or a regenerator, and focus fire on it.
- Perhaps there's a way to identify good terrain features nearby for particular activities, like a place to go to reload. I could have the NPC go there, but I'd have to figure out how they'd decide what's a good location.
- perhaps there's a way to get the NPC to remember the location of a vehicle they recently left, and treat that as a good "long distance" retreat target. Similarly maybe they can remember the last place they were safe without any zombies. for these, I'd need a fail state... how would the NPC know if they shouldn't go back there?

I would also be open to ideas around how NPCs might change their movement behaviour depending on how outnumbered they are, or around other situatoins. Should they act differently in the dark?

Another thing I could look into is having them recognize how many allies they have, and picking 'roles' on their own. Like, if one NPC has more armour, they might flag themselves as "the tank" and if one NPC is faster they might flag themselves as "the flanker". If I did that, how should I start adjusting their behaviours with the tools I have outlined?

Finally, what I'd like the absolute most is to hear what they're doing or not doing that drives you the most crazy, so I can see if I can fix that.

**Comment by DukePaulAtreid3s** [comment]
I would say that melee characters should stay in cover until the enemy is x tiles away from them (if said enemy has a gun), stick close to ranged buddies, and only engage enemies that are melee if they want to leave it

**Reply by I-am-Erk** [reply]
cross posting from our discord discussion to keep it recorded:
What I'd probably do for a first pass is something like:
1. check if player ally
2. If player ally, check if player has gun. if not player ally, check if nearby allies have guns.
3. If player ally, check combat engagement rules to see if instructed to stay close. If not, have at 'er. If yes, do not charge until enemy is approaching engagement range.
4. Otherwise, treat combat engagement rules as if you were instructed to stay close to nearby gun-toting allies using same algorithm.
I think we can actually ignore whether or not hte npc in question is using a melee weapon. Even if they have a shorter range gun they shouldn't charge in
Where this could possibly be refined is that if they haven't been isntructed to stay close, but they have a sword and you have an m16a1, then they should behave as if told to stay close unless specifically told to charge in
that way you don't have to micro them. If you've got a gun, your buddies with swords don't have to be reminded not to run into your engagement zone

**Comment by worm-girl** [comment]
I recently started in a stadium with an NPC pal and watched her repeatedly retreat deeper and deeper into the building, aggroing more enemies as she went, which inevitably got her killed. NPCs should be able to understand that this is the likely outcome when retreating to an unexplored area and only do it as a last resort or when they go into a total panic.

One specific thing an NPC might want to do if they recognize that they're in a dangerous overmap tile is to clear a single room and define it as a hiding place, then hang out in there, especially if there's darkness to take advantage of (unless they're nyctophobic). This behavior could even form the basis of future organic player/npc interactions. Finding an NPC hiding in such a place might prompt a dialog like "I'm trying to hide, can you help me escape?" or any number of other things that fit the scenario.

**Reply by I-am-Erk** [reply]
Having an NPC recognize an empty room and remember it as a retreat location is something I want to add, and will probably accomplish that. I need to figure out how to do that first, though.

**Comment by I-am-Erk** [comment]
Copying lazybuns comment from discord so I can keep track:
> Playing with npc follower now, with only ranged weapon
> feels like there is a rubber band tied to you and npc tries to kite as long as its not too far from you, then they commit to melee even without suitable weapon

> Problem with non-targeting of retreating and low danger targets still persists, mi-go's running away broke npc aggro so player had to finish them off all the time.
Overall behavior is reasonably predictable but i'd love to be able to change target assessment of npc somehow

**Reply by I-am-Erk** [reply]
It's possibly complex to change NPC behaviour to target those monsters that are harriers, but I think I can do it. It's definitely incorrect and it's caused by hit and run monsters "abusing" the Fleeing tag. What I can look into is trying to lower the priority of fleeing monsters but not drop it to zero.

I think the "rubber band" feeling is just me needing to add a clear_path call when NPCs are supposed to just move a square or two. I can add that easy

**Reply by worm-girl** [reply]
The hit and run code isn't super complex. Maybe monsters just need a new behavior flag for all instances where the monster is repositioning somehow, like "maneuvering" or something.

**Reply by I-am-Erk** [reply]
I solved it in an even easier way, actually. I just made hit and run monsters stay hostile while running.

**Comment by Spicyshadow** [comment]
You could implement ranged kiting for NPC´s. This could be as simple as a check to see if the speed differance between NPC and a threat. If it is big enough than the danger level of the enemy should drop significantly but the NPC should still keep a certain distance. Example: A NPC on a horse sees that the speed differance between it and a hulk is very large. Instead of running away like normal it stays much closer and starts shoting the hulk with arrows before repositioning again. This would need some additional checks too check for armor and regen too prevent a NPC from pelting rocks at a skeletal juggernaut. 

NPC´s could seek some form of rudamentary cover. Think if they are inrange of a enemy with a ranged attack than they try to break line of sight if possible while roughly remembering where the enemy was to prevent reastablising line of sight. This could be further expanded by having the NPC not move from thier position once they break line of sight to allow them to ambush the ranged enemy.

Implementing formations could be a very simple addition to the game. Just have the player make a grid where they place everyone and the NPC´s would try to they in that grid. Maybe make everyone move in lock step or force their turns to go after one another to prevent people from being caught out of position. This would give players and NPC´s advantages:

- In the form of buffs to reflect the advantages of fighting in a formation.
- A player could leverage the control they have over the formation to get tactical advantages. By placing NPC´s with reach weapons in the second to third row they can get more killing power at the front. It also has the advantage that any enemy that comes to close  is always directly fighting multiple NPC´s allowing you to partly stunlock a powerful enemy like a skeletal juggernaut becease 6 dudes are hitting it with halbards. You could also optimize the whole formation by presicely contoling who goes where with what equipment.

I would first implement formations for players. After a few months players will have developed a range of differant formations and strategies that you can than implement for NPC´s and perhaps monsters too use. Why design stuff yourself if you can trick the playersbase to do it for you?

**Comment by brokeboiflex** [comment]
Would be great if NPCs were enviroment aware, and using it to their advantage by luring enemies into traps, fires and ledges

**Comment by Polygonaldreamsgames** [comment]
It would be great if NPCs equipped with ranged weapons-especially rifles-tried to take up distant positions overlooking the target.
